File: packages/typescript/src/platforms/facebook.ts
/**
 * Facebook adapter.
 *
 * Notes:
 *   - Uses facebook.com (NOT m.facebook.com) for the rich composer.
 *   - Heavy fingerprint detection; recommend Browserbase residential proxies
 *     and a warm-up period for new accounts.
 */

import { BasePlatform } from "../core/BasePlatform.js";
import { resolveMfa } from "../core/mfa.js";
import type {
  CommentInput,
  CommentResult,
  Credentials,
  FollowInput,
  FollowResult,
  LikeInput,
  LikeResult,
  PostInput,
  PostResult,
} from "../core/types.js";

export class FacebookAdapter extends BasePlatform {
  protected loginUrl(): string {
    return "https://www.facebook.com/login/";
  }

  protected async performLogin(creds: Credentials): Promise<void> {
    const sh = await this.getStagehand();
    await sh.page.act(`Type "${creds.username}" into the email or phone field`);
    await sh.page.act(`Type "${creds.password}" into the password field`);
    await sh.page.act("Click the Log in button");
    await sh.page.waitForLoadState("networkidle").catch(() => {});

    const mfa = await sh.page.extract({
      instruction: "Is Facebook asking for a 2FA / login code? Return mfa=true|false plus hint.",
      schema: { mfa: "boolean", hint: "string" },
    }).catch(() => ({ mfa: false, hint: "" }));
    const m = mfa as { mfa: boolean; hint?: string };
    if (m.mfa) {
      const code = await resolveMfa(
        { description: "Facebook 2FA", channel: "unknown", hint: m.hint },
        creds,
        this.onMfaChallenge,
      );
      await sh.page.act(`Type "${code}" into the code field`);
      await sh.page.act("Click Continue");
      await sh.page.waitForLoadState("networkidle").catch(() => {});
    }

    // Dismiss "Trust this device?" / "Save browser?" prompts.
    await sh.page.act("If a 'Trust this device?' or 'Save browser?' prompt appears, click Continue or Trust this device").catch(() => undefined);
  }

  async isLoggedIn(): Promise<boolean> {
    const sh = await this.getStagehand();
    if (!sh.page.url().includes("facebook.com")) {
      await sh.page.goto("https://www.facebook.com/").catch(() => {});
    }
    const r = await sh.page.extract({
      instruction:
        "Is the user logged into Facebook? Look for the blue header bar with the user's account avatar and a 'What's on your mind?' composer. Return loggedIn=true|false.",
      schema: { loggedIn: "boolean" },
    }).catch(() => ({ loggedIn: false }));
    return Boolean((r as { loggedIn: boolean }).loggedIn);
  }

  async post(input: PostInput): Promise<PostResult> {
    const sh = await this.getStagehand();
    await sh.page.goto("https://www.facebook.com/");
    await sh.page.act("Click the 'What's on your mind?' composer at the top of the feed");

    if (input.media?.length) {
      const chooserPromise = sh.page.waitForEvent("filechooser");
      await sh.page.act("Click the 'Photo/video' button in the post dialog");
      const chooser = await chooserPromise;
      await chooser.setFiles(input.media);
    }

    if (input.text) {
      const caption = this.composeCaption(input);
      await sh.page.act(`Type the following into the post text area: ${JSON.stringify(caption)}`);
    }

    await sh.page.act("Click the 'Post' button at the bottom of the dialog");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now() };
  }

  private composeCaption(input: PostInput): string {
    let caption = input.text ?? "";
    if (input.hashtags?.length) {
      caption += "\n\n" + input.hashtags.map((h) => `#${h.replace(/^#/, "")}`).join(" ");
    }
    return caption;
  }

  async comment(input: CommentInput): Promise<CommentResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the 'Write a comment...' input under the post");
    await sh.page.act(`Type the comment: ${JSON.stringify(input.text)}`);
    await sh.page.act("Press Enter to submit the comment");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url };
  }

  async like(input: LikeInput): Promise<LikeResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the 'Like' button (thumbs up) under the post");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url, liked: true };
  }

  async unlike(input: LikeInput): Promise<LikeResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the (currently-active blue) Like button under the post to unlike it");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url, liked: false };
  }

  async follow(input: FollowInput): Promise<FollowResult> {
    const sh = await this.getStagehand();
    const handle = input.username.replace(/^@/, "");
    await sh.page.goto(`https://www.facebook.com/${handle}`);
    await sh.page.act("Click the 'Follow' button on the profile header. If you see 'Add Friend' instead, click that.");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), following: true };
  }

  async unfollow(input: FollowInput): Promise<FollowResult> {
    const sh = await this.getStagehand();
    const handle = input.username.replace(/^@/, "");
    await sh.page.goto(`https://www.facebook.com/${handle}`);
    await sh.page.act("Click the 'Following' button, then click 'Unfollow' in the menu");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), following: false };
  }
}

File: packages/python/stagehand_social/platforms/facebook.py
"""Facebook adapter — Python mirror of platforms/facebook.ts."""

from __future__ import annotations

import json

from stagehand_social.core.base_platform import BasePlatform
from stagehand_social.core.mfa import resolve_mfa
from stagehand_social.core.types import (
    CommentInput,
    CommentResult,
    Credentials,
    FollowInput,
    FollowResult,
    LikeInput,
    LikeResult,
    MfaChallenge,
    PostInput,
    PostResult,
)


class FacebookAdapter(BasePlatform):
    def login_url(self) -> str:
        return "https://www.facebook.com/login/"

    async def perform_login(self, creds: Credentials) -> None:
        sh = await self.get_stagehand()
        await sh.page.act(f'Type "{creds.username}" into the email or phone field')
        await sh.page.act(f'Type "{creds.password}" into the password field')
        await sh.page.act("Click the Log in button")
        try:
            await sh.page.wait_for_load_state("networkidle")
        except Exception:
            pass

        try:
            mfa = await sh.page.extract(
                instruction="Is Facebook asking for a 2FA / login code? Return mfa=true|false plus hint.",
                schema={"mfa": "boolean", "hint": "string"},
            )
        except Exception:
            mfa = {"mfa": False, "hint": ""}
        if isinstance(mfa, dict) and mfa.get("mfa"):
            code = await resolve_mfa(
                MfaChallenge(description="Facebook 2FA", channel="unknown", hint=mfa.get("hint")),
                creds,
                self.on_mfa_challenge,
            )
            await sh.page.act(f'Type "{code}" into the code field')
            await sh.page.act("Click Continue")

        try:
            await sh.page.act(
                "If a 'Trust this device?' or 'Save browser?' prompt appears, click Continue or Trust this device"
            )
        except Exception:
            pass

    async def is_logged_in(self) -> bool:
        sh = await self.get_stagehand()
        if "facebook.com" not in sh.page.url:
            try:
                await sh.page.goto("https://www.facebook.com/")
            except Exception:
                pass
        try:
            r = await sh.page.extract(
                instruction=(
                    "Is the user logged into Facebook? Look for the blue header bar with avatar "
                    "and a 'What's on your mind?' composer. Return loggedIn=true|false."
                ),
                schema={"loggedIn": "boolean"},
            )
            return bool(isinstance(r, dict) and r.get("loggedIn"))
        except Exception:
            return False

    async def post(self, input: PostInput) -> PostResult:
        sh = await self.get_stagehand()
        await sh.page.goto("https://www.facebook.com/")
        await sh.page.act("Click the 'What's on your mind?' composer at the top of the feed")

        if input.media:
            async with sh.page.expect_file_chooser() as fc_info:
                await sh.page.act("Click the 'Photo/video' button in the post dialog")
            chooser = await fc_info.value
            await chooser.set_files(input.media)

        if input.text:
            caption = self._compose_caption(input)
            await sh.page.act(f"Type the following into the post text area: {json.dumps(caption)}")

        await sh.page.act("Click the 'Post' button at the bottom of the dialog")
        return PostResult(success=True, at=self._now())

    def _compose_caption(self, input: PostInput) -> str:
        caption = input.text or ""
        if input.hashtags:
            caption += "\n\n" + " ".join(f"#{h.lstrip('#')}" for h in input.hashtags)
        return caption

    async def comment(self, input: CommentInput) -> CommentResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the 'Write a comment...' input under the post")
        await sh.page.act(f"Type the comment: {json.dumps(input.text)}")
        await sh.page.act("Press Enter to submit the comment")
        return CommentResult(success=True, at=self._now(), url=input.url)

    async def like(self, input: LikeInput) -> LikeResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the 'Like' button (thumbs up) under the post")
        return LikeResult(success=True, at=self._now(), url=input.url, liked=True)

    async def unlike(self, input: LikeInput) -> LikeResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the (currently-active blue) Like button under the post to unlike it")
        return LikeResult(success=True, at=self._now(), url=input.url, liked=False)

    async def follow(self, input: FollowInput) -> FollowResult:
        sh = await self.get_stagehand()
        handle = input.username.lstrip("@")
        await sh.page.goto(f"https://www.facebook.com/{handle}")
        await sh.page.act(
            "Click the 'Follow' button on the profile header. If only 'Add Friend' is visible, click that."
        )
        return FollowResult(success=True, at=self._now(), following=True)

    async def unfollow(self, input: FollowInput) -> FollowResult:
        sh = await self.get_stagehand()
        handle = input.username.lstrip("@")
        await sh.page.goto(f"https://www.facebook.com/{handle}")
        await sh.page.act("Click the 'Following' button, then click 'Unfollow' in the menu")
        return FollowResult(success=True, at=self._now(), following=False)
